src/vmm/src/guest_config/x86_64/mod.rs

```diff
@@ -1,17 +1,376 @@
+#[derive(Debug, Clone, PartialEq)]
 pub struct X86_64CpuConfiguration {
     /// CPUID configuration
     pub cpuid: Cpuid,
     /// Register values as a key pair for model specific registers
     /// Key: MSR address
     /// Value: MSR value
     pub msrs: HashMap<u32, u64>,
+
+    /// Set of supported MSRs
+    pub msr_list: HashSet<u32>,
+
+    /// Architectural MSPs required for boot
+    pub msr_boot_entries: Vec<kvm_msr_entry>,
+}
+
+impl X86_64CpuConfiguration {
+    /// Craetes new CpuConfig with default values
+    pub fn new(vm: &Vm) -> Result<Self, Error> {
+        let supported_cpuid = vm.supported_cpuid().clone();
+        let supported_cpuid = Cpuid::try_from(RawCpuid::from(supported_cpuid)).unwrap();
+        let msr_list = vm.supported_msrs().as_slice().iter().copied().collect();
+
+        let msr_boot_entries = create_boot_msr_entries();
+
+        Ok(Self {
+            cpuid: supported_cpuid,
+            msrs: Default::default(),
+            msr_list,
+            msr_boot_entries,
+        })
+    }
+
+    /// Craetes new CpuConfig with cpu template changes applied
+    pub fn from_cpu_template(vm: &Vm, template: &CpuTemplateType) -> Result<Self, Error> {
+        let supported_cpuid = vm.supported_cpuid().clone();
+        let mut supported_cpuid = Cpuid::try_from(RawCpuid::from(supported_cpuid)).unwrap();
+        let mut msr_list = vm.supported_msrs().as_slice().iter().copied().collect();
+
+        // Initialize some architectural MSRs that will be set for boot.
+        let mut msr_boot_entries = create_boot_msr_entries();
+
+        match template {
+            CpuTemplateType::Custom(template) => {
+                // TODO QUERY ADDITIONAL MSRS
+                let config = Self {
+                    cpuid: supported_cpuid,
+                    msrs: Default::default(),
+                    msr_list,
+                    msr_boot_entries,
+                };
+
+                Self::with_template(config, template)
+            }
+            CpuTemplateType::Static(template) => {
+                // If a template is specified, get the CPUID template, else use `cpuid`.
+                let config_cpuid = match template {
+                    CpuFeaturesTemplate::C3 => cpuid_templates::c3::template(),
+                    CpuFeaturesTemplate::T2 => cpuid_templates::t2::template(),
+                    CpuFeaturesTemplate::T2S => cpuid_templates::t2s::template(),
+                    CpuFeaturesTemplate::T2CL => cpuid_templates::t2cl::template(),
+                    CpuFeaturesTemplate::T2A => cpuid_templates::t2a::template(),
+                    CpuFeaturesTemplate::None | CpuFeaturesTemplate::Custom => {
+                        unreachable!("None and Custom states are invalid")
+                    }
+                };
+
+                // Include leaves from host that are not present in CPUID template.
+                supported_cpuid = config_cpuid
+                    .include_leaves_from(supported_cpuid)
```

Comment:
I *suspect* the inclusion of the leaves from the supported config needs to happen after [normalization in vcpu.configure()](https://github.com/firecracker-microvm/firecracker/pull/3475/files/faafb85f665efc7dfdadf288fc23da91bd33b6c9#diff-e1ead8e51789d6db11155356a0b6c2b85f76f4648e5cd5ef882b7cca6b710a52L295). Have a quick look at [the build error](https://buildkite.com/firecracker/firecracker-pr/builds/1800#018708e6-ff0f-495b-9b88-b37328d17e88) to try reproduce.

Reply:
Well, `inclusion` only fails if the `config_cpuid` and `supported_cpuid` are from different vendors.
The test that fails expects the vm to fail to start.
```python
fail_msg = "Internal error while starting microVM: Error configuring the vcpu for boot:"
assert test_microvm.api_session.is_status_bad_request(response.status_code)
assert fail_msg in response.text
```
It is just the error message I need to fix